File: Python implementation/gpa.py
import numpy as np
import matplotlib.pyplot as plt

def CPA(X, Y, scale=True, translation=True):
    '''
    Procrustes analysis with optional scaling and translation to align X to Y.
    Both X and Y are arrays of shape (m, n) where m and n are arbitrary.
    Returns: aligned_X, rotation_matrix, scale, translation
    '''

    if translation:
        X_mean = X.mean(axis=0)
        Y_mean = Y.mean(axis=0)
        X0 = X - X_mean
        Y0 = Y - Y_mean
    else:
        X_mean = np.zeros(X.shape[1], dtype=X.dtype)
        Y_mean = np.zeros(Y.shape[1], dtype=Y.dtype)
        X0 = X
        Y0 = Y

    U, _, Vt = np.linalg.svd(X0.T @ Y0)
    R = U @ Vt

    if np.linalg.det(R) < 0:
        Vt[-1, :] *= -1
        R = U @ Vt

    s = 1
    if scale:
        s = np.trace((X0 @ R).T @ Y0) / np.trace(X0.T @ X0)

    aligned_X = s * X0 @ R + Y_mean

    return aligned_X, R, s, Y_mean - s * X_mean @ R


def GPA(X_list, scale=True, epsilon=1e-4):
    '''
    Generalized Procrustes analysis with optional scaling to align X to Y.
    Both X and Y are arrays of shape (m, n) where m and n are arbitrary.
    Returns: Y_consensus, aligned_X_list, scaling_factors
    '''

    m = len(X_list)

    # STEP 1
    X_list = [np.array(X.copy()) for X in X_list]
    for i in range(m):
        X_mean = X_list[i].mean(axis=0)
        X_list[i] = X_list[i] - X_mean

    fro_sum = 0
    for i in range(m):
        fro_sum += np.linalg.norm(X_list[i], 'fro')**2

    var_lambda = np.sqrt(m / fro_sum)

    for i in range(m):
        X_list[i] = X_list[i] * var_lambda

    Y_consensus = X_list[0]

    # STEP 2
    for i in range(1, m):
        X_list[i] = CPA(X_list[i], Y_consensus, scale=False, translation=False)[0]

    Y_consensus = np.sum(np.stack(X_list, axis=0), axis=0) / m
    Sr = m * (1 - np.trace(Y_consensus @ Y_consensus.T))
    print(Sr)
    scaling_factors = np.array([var_lambda for _ in range(m)])

    # STEP 3
    while True:
        for i in range(m):
            X_list[i] = CPA(X_list[i], Y_consensus, scale=False, translation=False)[0]

        # STEP 4 (nothing is done in the optimized version)
        # STEP 5
        if scale:
            for i in range(m):
                change_factor = np.sqrt(np.trace(X_list[i] @ Y_consensus.T) / \
                    (np.trace(X_list[i] @ X_list[i].T) * np.trace(Y_consensus @ Y_consensus.T)))
                s_i_star = scaling_factors[i] * change_factor
                X_list[i] *= s_i_star / scaling_factors[i]
                scaling_factors[i] = s_i_star

        Y_consensus_star_star = np.sum(np.stack(X_list, axis=0), axis=0) / m
        Sr_star_star = Sr - m * np.trace(Y_consensus_star_star @ Y_consensus_star_star.T - Y_consensus @ Y_consensus.T)
        print(Sr_star_star)
        Y_consensus = Y_consensus_star_star

        # STEP 6
        if Sr - Sr_star_star < epsilon:
            break
        else:
            Sr = Sr_star_star
    
    # STEP 7
    return Y_consensus, X_list, scaling_factors

File: testing_code/gpa.cpp
#include <iostream>
#include <Eigen/Dense>
#include <tuple>
#include <vector>
#include <cmath>
#include <fstream>
#include <iomanip>

using namespace Eigen;

std::tuple<MatrixXd, MatrixXd, double, MatrixXd> CPA(
    const MatrixXd& X, 
    const MatrixXd& Y, 
    bool scale = true, 
    bool translation = true)
{
    MatrixXd X0, Y0;
    VectorXd X_mean, Y_mean;

    if (translation) {
        // Compute column-wise mean
        X_mean = X.colwise().mean();
        Y_mean = Y.colwise().mean();

        // Center matrices
        X0 = X.rowwise() - X_mean.transpose();
        Y0 = Y.rowwise() - Y_mean.transpose();
    } else {
        X_mean = VectorXd::Zero(X.cols());
        Y_mean = VectorXd::Zero(Y.cols());
        X0 = X;
        Y0 = Y;
    }

    // Singular value decomposition (SVD) of X0^T * Y0
    BDCSVD<MatrixXd, ComputeFullU | ComputeFullV> svd(X0.transpose() * Y0);
    MatrixXd U = svd.matrixU();
    MatrixXd V = svd.matrixV();
    MatrixXd R = U * V.transpose();

    // Reflection check
    if (R.determinant() < 0) {
        V.col(V.cols() - 1) *= -1;
        R = U * V.transpose();
    }

    // Scaling if needed
    double s = 1.0;
    if (scale) {
        s = ((X0 * R).array() * Y0.array()).sum() / X0.squaredNorm();
    }

    // Producing the aligned X
    MatrixXd aligned_X = s * X0 * R;
    aligned_X.rowwise() += Y_mean.transpose();

    // Translation component (optional)
    MatrixXd translation_vec = Y_mean.transpose() - s * (X_mean.transpose() * R);

    return {aligned_X, R, s, translation_vec};
}


std::tuple<MatrixXd, std::vector<MatrixXd>, std::vector<double>> GPA(
    const std::vector<MatrixXd>& X_list_input, 
    bool scale = true, 
    double epsilon = 1e-4)
{
    int m = X_list_input.size();
    std::vector<MatrixXd> X_list(m);

    // STEP 1: Copy and center the matrices
    double fro_sum = 0;
    for (int i = 0; i < m; ++i) {
        X_list[i] = X_list_input[i];
        VectorXd X_mean = X_list[i].colwise().mean();
        X_list[i].rowwise() -= X_mean.transpose();
        fro_sum += X_list[i].squaredNorm();
    }

    double var_lambda = std::sqrt(static_cast<double>(m) / fro_sum);
    std::vector<double> scaling_factors(m, var_lambda);

    for (int i = 0; i < m; ++i) {
        X_list[i] *= var_lambda;
    }

    MatrixXd Y_consensus = X_list[0];

    // STEP 2: Align the initial matrices to the consensus
    for (int i = 1; i < m; ++i) {
        X_list[i] = std::get<0>(CPA(X_list[i], Y_consensus, false, false));
    }

    Y_consensus = MatrixXd::Zero(Y_consensus.rows(), Y_consensus.cols());
    for (int i = 0; i < m; ++i) {
        Y_consensus += X_list[i];
    }
    Y_consensus /= m;

    double Sr = m * (1 - Y_consensus.squaredNorm());

    // STEP 3: Iterative alignment
    while (true) {
        // Align all the matrices to the current consensus
        for (int i = 0; i < m; ++i) {
            X_list[i] = std::get<0>(CPA(X_list[i], Y_consensus, false, false));
        }

        // STEP 4: Skipped in this opimized version
        // See "Практические аспекты"

        // STEP 5: Optional scaling
        if (scale) {
            for (int i = 0; i < m; ++i) {
                double numerator = (X_list[i].array() * Y_consensus.array()).sum();
                double denominator = X_list[i].squaredNorm() * Y_consensus.squaredNorm();
                double change_factor = std::sqrt(numerator / denominator);
                double s_i_star = scaling_factors[i] * change_factor;
                X_list[i] *= s_i_star / scaling_factors[i];
                scaling_factors[i] = s_i_star;
            }
        }

        // Update consensus
        MatrixXd Y_consensus_new = MatrixXd::Zero(Y_consensus.rows(), Y_consensus.cols());
        for (int i = 0; i < m; ++i) {
            Y_consensus_new += X_list[i];
        }
        Y_consensus_new /= m;

        double Sr_new = Sr - m * (Y_consensus_new.squaredNorm() - Y_consensus.squaredNorm());
        Y_consensus = Y_consensus_new;

        // STEP 6: Check for the termination condition
        if (Sr - Sr_new < epsilon) break;

        Sr = Sr_new;
    }

    // STEP 7: Exit the procedure
    return {Y_consensus, X_list, scaling_factors};
}


Eigen::MatrixXd readMatrix(const std::string &filename, int rows, int cols) {
    Eigen::MatrixXd mat(rows, cols);
    std::ifstream in(filename);
    if(!in) {
        std::cerr << "Cannot open file " << filename << std::endl;
        exit(1);
    }
    for(int i=0;i<rows;++i)
        for(int j=0;j<cols;++j)
            in >> mat(i,j);
    return mat;
}

void writeMatrix(const Eigen::MatrixXd& mat, const std::string& filename) {
    std::ofstream out(filename);
    if (!out) {
        std::cerr << "Cannot open file " << filename << " for writing.\n";
        return;
    }
    out << std::fixed << std::setprecision(8); // optional: match input precision
    for (int i = 0; i < mat.rows(); ++i) {
        for (int j = 0; j < mat.cols(); ++j) {
            out << mat(i, j);
            if (j < mat.cols() - 1) out << " ";
        }
        out << "\n";
    }
    out.close();
}

int main() {

    Eigen::MatrixXd X0 = readMatrix("X_0.txt", 17, 2);
    Eigen::MatrixXd X1 = readMatrix("X_1.txt", 17, 2);
    Eigen::MatrixXd X2 = readMatrix("X_2.txt", 17, 2);
    Eigen::MatrixXd X3 = readMatrix("X_3.txt", 17, 2);
    Eigen::MatrixXd X4 = readMatrix("X_4.txt", 17, 2);
    Eigen::MatrixXd X5 = readMatrix("X_5.txt", 17, 2);
    std::vector<Eigen::MatrixXd> X_list = {X0, X1, X2, X3, X4, X5};

    /*
    // std::cout << "X0 =\n" << X0 << std::endl;
    auto [aligned_X, R, s, translation_vec] = CPA(X4, X5, true);
    std::cout << "Aligned one to another:\n" << aligned_X << "\n\n";
    std::cout << "Scale factor s:\n" << s << "\n\n";
    std::cout << "Rotation matrix R:\n" << R << "\n\n";
    */

    auto [Y_consensus, aligned_X_list, scaling_factors] = GPA(X_list, true, 1e-15);
    // Assume aligned_X_list has 6 matrices
    for (size_t i = 0; i < aligned_X_list.size(); ++i) {
        std::string filename = "aligned_X_" + std::to_string(i) + ".txt";
        writeMatrix(aligned_X_list[i], filename);
    }

    // Write consensus
    writeMatrix(Y_consensus, "Y_consensus.txt");

    for (int i = 0; i < scaling_factors.size(); ++i) {
        std::cout << scaling_factors[i] << ' ';
    }

    return 0;

}
